### PR TITLE
docs!: generate documentation with compiled examples

### PR DESCRIPTION
**Pull request by jeromevdl**
**Issue #, if available:**

## Description of changes:

- Created submodules (per layer)
- Created one README per submodule, containing the documentation for all the constructs of this submodule
- Created examples (`*.lit.ts`) that will be compiled by the build process
- Referenced these examples from the README
- Created a script that takes the output of `jsii` (`.jsii` file) and `jsii-rosetta` (`.jsii.tabl.json` file) to generate one MD per construct containing the documentation from the README + the code (ts and python) from the examples and put them in the website structure.
- Modified the website pages to import the generated pages

**Checklist**

 

* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/) (`fix: `, `feat: `, `docs: `, ...)

## Breaking change checklist

submodules => need to change the imports (ok as we are still in RC)

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by vgkowski — CHANGES_REQUESTED** [review]
We need to document the process for adding a construct in the doc

**Review by vgkowski — DISMISSED** [review]

**Comment by vgkowski** [comment]
We will document the process to add a construct in the doc in another PR

**Comment by jeromevdl** [comment]
> We need to document the process for adding a construct in the doc

as discussed, I will do it in another PR so that we can merge this one quickly and avoid other conflicts

File: .projenrc.ts [hunk 1]
```diff
@@ -152,8 +152,22 @@ const fwkProject = new awscdk.AwsCdkConstructLibrary({
       target: "ES2021"
     }
   },
+
+  eslintOptions: {
+    dirs: ['src'],
```
Comment (balabad) [inline .projenrc.ts:157]: Do we also want to enable it for tests?

File: .projenrc.ts [hunk 2]
```diff
@@ -232,5 +248,6 @@ rootProject.addTask('test:e2e', {
 rootProject.addTask('release',  {
   description: 'Release project',
 })
-
+rootProject.addGitIgnore(".jsii.tabl.json");
+rootProject.addPackageIgnore("!*.lit.ts");
```
Comment (balabad) [inline .projenrc.ts:252]: This one is already in the `fwkProject`. Do we need it at the top level or just inside of the framework project?

File: framework/generate_doc.mjs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import data from './.jsii.tabl.json' assert { type: 'json' };
+
+// select only snippets for readme files
+const snippets = Object.fromEntries(
+    Object.entries(data.snippets)
+        .filter(([_,v]) => v.location.api.api === 'moduleReadme')
+);
+
+const linksMapping = {
+    '../storage/README.md#datalakestorage' : '../02-Storage/03-data-lake-storage.mdx',
```
Comment (balabad) [inline framework/generate_doc.mjs:11]: Any way to auto-generate this mapping? Or, how to ensure we can properly maintain it manually?

File: framework/generate_doc.mjs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        previousStart = 0;
+        codeLength = 0;
+    }
+    const file = `src/${submodule}/README.md`;
+
+    // do not read file on each snippet
+    if (file !== currentMDFile) {
+        currentMDFile = file;
+        const buffer= fs.readFileSync(file);
+        const str= buffer.toString();
```
Comment (balabad) [inline framework/generate_doc.mjs:40]:
```suggestion
        const buffer = fs.readFileSync(file);
        const str = buffer.toString();
```

File: framework/generate_doc.mjs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        lines = str.split("\n");
+    }
+
+    let line = snippet.location.field.line;
+    let realLine = line - codeLength;
+
+    // jsii does not count comments, we must add lines for each one
+    // let numberOfComments = lines.slice(0, realLine).filter(elt => elt.startsWith("[//]: # ")).length;
+    // realLine += numberOfComments;
```
Comment (balabad) [inline framework/generate_doc.mjs:49]: Do we need this commented code?

File: framework/generate_doc.mjs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+const titleRegEx = new RegExp('\\[\\/\\/\]: # \\((.*)\\)\\n', 'gm');
+
+for (let module in fullReadme) {
+    const moduleReadme = fullReadme[module];
+    const constructReadmes = moduleReadme.split(titleRegEx);
+    let filename = '';
+    for (let i = 0; i < constructReadmes.length; i++) {
+        let constructReadme = constructReadmes[i];
+        if (constructReadme.startsWith(module)) {
+            filename = '../website/docs/constructs/library/generated/_' + constructReadme.replace('.', '-') + '.mdx';
```
Comment (balabad) [inline framework/generate_doc.mjs:91]: Hard-coded paths... Maybe parameterize the script or at least define as `const` in the beginning of the file?

File: framework/generate_doc.mjs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        } else if (constructReadme.length !== 0 && filename.length !== 0) {
+            let constructReadmeLines = constructReadme.split('\n');
+            constructReadmeLines.shift(); // remove title
+            constructReadme = constructReadmeLines.join('\n');
+            constructReadme = `[//]: # (This file is generated, do not modify directly, update the README.md in framework/src/${module})
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+${constructReadme}`;
+
+            constructReadme = constructReadme.replaceAll("website/static/img", "../static/img"); // change image path
```
Comment (balabad) [inline framework/generate_doc.mjs:102]: Hard-coded path, see comment above.

File: .projenrc.ts (hunk 1 above)
Comment (jeromevdl) [reply .projenrc.ts:157]: I initially did it but it modifies the [tasks.json](https://github.com/awslabs/aws-data-solutions-framework/blob/5b635bdd23014aba24dff48d51998d5f0d3e2d7b/framework/.projen/tasks.json#L77) and add `test` twice...

Comment (jeromevdl) [reply .projenrc.ts:157]: as you can see test is already there...

File: .projenrc.ts (hunk 2 above)
Comment (jeromevdl) [reply .projenrc.ts:252]: removing from the root.

File: framework/generate_doc.mjs (hunk 3 above)
Comment (jeromevdl) [reply framework/generate_doc.mjs:11]: no way, that's the only thing I could not automate... but if links are not good, docusaurus fails so there is at least a guardrail

File: framework/generate_doc.mjs (hunk 5 above)
Comment (jeromevdl) [reply framework/generate_doc.mjs:49]: nope

**Review by balabad — APPROVED** [review]

**Pull request merged**